TESTS/mbed_drivers/ticker/main.cpp

```diff
@@ -14,58 +14,95 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
-
-/*
- * Tests is to measure the accuracy of Ticker over a period of time
- *
- *
- * 1) DUT would start to update callback_trigger_count every milli sec, in 2x callback we use 2 tickers
- *    to update the count alternatively.
- * 2) Host would query what is current count base_time, Device responds by the callback_trigger_count
- * 3) Host after waiting for measurement stretch. It will query for device time again final_time.
- * 4) Host computes the drift considering base_time, final_time, transport delay and measurement stretch
- * 5) Finally host send the results back to device pass/fail based on tolerance.
- * 6) More details on tests can be found in timing_drift_auto.py
- */
-
 #include "mbed.h"
 #include "greentea-client/test_env.h"
 #include "utest/utest.h"
 #include "unity/unity.h"
 
-using namespace utest::v1;
+
+using utest::v1::Case;
 
 #define ONE_MILLI_SEC 1000
+#define TICKER_COUNT 16
+#define MULTI_TICKER_TIME_MS 100
 volatile uint32_t callback_trigger_count = 0;
 static const int test_timeout = 240;
 static const int total_ticks = 10;
 
+extern uint32_t SystemCoreClock; /* CPU clock */
+#define TOLERANCE_FACTOR 15000.0f
+#define US_FACTOR 1000000.0f
+#define TOLERANCE ((TOLERANCE_FACTOR / (float)SystemCoreClock) * US_FACTOR)
+/* Tolerance in microseconds
+
+   Assume below code
+
+   ...
+   Ticker ticker;
+   Timer gtimer;
+   Semaphore sem(0, 1);
+
+   void stop_gtimer_set_flag(Semaphore *sem)
+   {
+       gtimer.stop();
+       stop_gtimer_set_flag->relase();
+   }
+
+   gtimer.start();
+   ticker.attach_us(callback(stop_gtimer_set_flag, &sem), 100000);
+   sem.wait();
+   int time_diff = gtimer.read_us();
+   TEST_ASSERT_UINT32_WITHIN((uint32_t)TOLERANCE, 100000, time_diff);
+   ...
+
+   For 100000[us] callback for NUCLEO_F070RB we get time_diff: 100060[us]
+   The extra 60[us] is from time measure functions and attach_us calls
+   Let's call this extra time X, then the measured time is 100000 + X [us]
+
+   To take in to account this extra time we introduces TOLERANCE value based on CPU clock (speed)
+   TOLERANCE_FACTOR was chosen experimentally in a way to give TOLERANCE equal roughly 5*X
+
+   e.g.
+   For K64F          TOLERANCE = (15000 / 120000000) * 1000000 =  125[us]
+   For NUCLEO_F070RB TOLERANCE = (15000 /  48000000) * 1000000 = 312.5[us]
+   For NRF51_DK      TOLERANCE = (15000 /  16000000) * 1000000 = 937.5[us]
+
+ */
+
+
+volatile bool ticker_callback_flag;
+volatile uint32_t multi_counter;
```

Comment:
Ideally those variables should be accessed atomically, volatile does not guarantee atomicity.

Reply:
you mean increment/change atomically ?

Reply:
Yes, use atomic operations we provide

Reply:
there is no function: `core_util_atomic_incr_u32(volatile uint32_t *valuePtr` so we have to cast to remove volatile: `core_util_atomic_incr_u32((uint32_t*)&multi_counter, 1);`

Other option is to use just `core_util_critical_section` to protect `multi_counter` change 
O mayby we should have `volatile` versions of `core_util_atomic_` functions ?
@0xc0170 what do you think ?

Reply:
`volatile` should not be required for atomic access because the memory order we use is similar to `std::memory_order_seq_cst` so data access should not be reordered.